org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/refactoring/impl/RefactoringUpdateAcceptor.java

```diff
@@ -50,23 +50,31 @@ public RefactoringUpdateAcceptor(IRefactoringDocument.Provider refactoringDocume
 	@Override
 	public void accept(URI resourceURI, TextEdit textEdit) {
 		IRefactoringDocument document = getDocument(resourceURI);
-		document2textEdits.put(document, textEdit);
+		if (document != null) {
+			document2textEdits.put(document, textEdit);
+		}
 	}
 
 	@Override
 	public void accept(URI resourceURI, Change change) {
 		IRefactoringDocument document = getDocument(resourceURI);
```

Comment:
This looks a little weird to me in general. How can I create a change without a document / resource?

Reply:
Looking at DefaultRefactoringDocumentProvider.get() (where the null is returned), it can only happen when no storage can be found for the given resourceUri.

Could this happen if you start a renaming an element, then remove the physical file from an external source and then continue the refactor?

Reply:
@szarnekow Went a little deeper: the document is null if for the given uri:
* there is no file
* the file is not editable

Reply:
> the file is not editable

That would be an attempt to refactor something, that is from a jar file?

Reply:
It doesn't mention if file is java.nio.file, eclipse file or physical file. But maybe this happens when you start refactoring and open the preview. Stay on the preview page for several minutes. Meanwhile, outside of Eclipse, change permissions of the file, or remove it. This should be testable so I'll try if I can reproduce that behavior.

Reply:
Tried changing permissions, got a dialog saying I had a readonly file and if I wanted to make it writable, said no, got a new stacktrace:
```
org.eclipse.swt.SWTException: Invalid thread access
	at org.eclipse.swt.SWT.error(SWT.java:4595)
	at org.eclipse.swt.SWT.error(SWT.java:4510)
	at org.eclipse.swt.SWT.error(SWT.java:4481)
	at org.eclipse.swt.widgets.Widget.error(Widget.java:506)
	at org.eclipse.swt.widgets.Shell.<init>(Shell.java:273)
	at org.eclipse.swt.widgets.Shell.<init>(Shell.java:370)
	at org.eclipse.jface.window.Window.createShell(Window.java:485)
	at org.eclipse.jface.window.Window.create(Window.java:428)
	at org.eclipse.jface.dialogs.Dialog.create(Dialog.java:1099)
	at org.eclipse.jface.window.Window.open(Window.java:786)
	at org.eclipse.jface.dialogs.ErrorDialog.open(ErrorDialog.java:348)
	at org.eclipse.jface.dialogs.ErrorDialog.openError(ErrorDialog.java:412)
	at org.eclipse.jface.dialogs.ErrorDialog.openError(ErrorDialog.java:377)
	at org.eclipse.ui.texteditor.AbstractTextEditor.openSaveErrorDialog(AbstractTextEditor.java:5090)
	at org.eclipse.ui.texteditor.AbstractDecoratedTextEditor.openSaveErrorDialog(AbstractDecoratedTextEditor.java:1645)
	at org.eclipse.ui.texteditor.AbstractTextEditor.handleExceptionOnSave(AbstractTextEditor.java:5062)
	at org.eclipse.ui.texteditor.AbstractTextEditor.performSave(AbstractTextEditor.java:5012)
	at org.eclipse.ui.texteditor.AbstractTextEditor.doSave(AbstractTextEditor.java:4806)
	at org.eclipse.xtext.ui.editor.XtextEditor.doSave(XtextEditor.java:318)
	at org.eclipse.xtext.ui.refactoring.impl.EditorDocumentChange.commit(EditorDocumentChange.java:90)
	at org.eclipse.ltk.core.refactoring.TextChange.perform(TextChange.java:242)
	at org.eclipse.ltk.core.refactoring.CompositeChange.perform(CompositeChange.java:283)
	at org.eclipse.ltk.core.refactoring.CompositeChange.perform(CompositeChange.java:283)
	at org.eclipse.ltk.core.refactoring.CompositeChange.perform(CompositeChange.java:283)
	at org.eclipse.ltk.core.refactoring.PerformChangeOperation$1.run(PerformChangeOperation.java:260)
	at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2292)
	at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2319)
	at org.eclipse.ltk.core.refactoring.PerformChangeOperation.executeChange(PerformChangeOperation.java:308)
	at org.eclipse.ltk.internal.ui.refactoring.UIPerformChangeOperation.executeChange(UIPerformChangeOperation.java:99)
	at org.eclipse.ltk.core.refactoring.PerformChangeOperation.run(PerformChangeOperation.java:224)
	at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2292)
	at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2319)
	at org.eclipse.ltk.internal.ui.refactoring.WorkbenchRunnableAdapter.run(WorkbenchRunnableAdapter.java:89)
	at org.eclipse.jface.operation.ModalContext$ModalContextThread.run(ModalContext.java:122)
```

Reply:
Same thing for removing files, can't reproduce. Tried renaming a class from a .class file in a jar as well, didn't work